File: EXAMPLES/rgpio/28BYJ_48.c
/*
28BYJ_48.c
2020-11-21
Public Domain

http://abyz.me.uk/lg/rgpio.html

gcc -Wall -o 28BYJ_48 28BYJ_48.c -lrgpio

./28BYJ_48 [chip] gpio1 gpio2 gpio3 gpio4

E.g.

./28BYJ_48 20 21 22 23 # gpiochip=0 gpio1=20 gpio2=21 gpio3=22 gpio4=23

./28BYJ_48 2 7 5 11 3  # gpiochip=2 gpio1=7 gpio2=5 gpio3=11 gpio4=3
*/

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <inttypes.h>

#include <lgpio.h>
#include <rgpio.h>

int coils[] = {7, 3, 11, 9, 13, 12, 14, 6};

int forward(int *pos)
{
   if (++(*pos) > 7) *pos = 0;
   return coils[*pos];
}

int backward(int *pos)
{
   if (--(*pos) < 0) *pos = 7;
   return coils[*pos];
}

int main(int argc, char *argv[])
{
   int sbc, h, err, i;
   int pos=0;
   int chip = 0;
   int GPIO[4];
   int ZEROS[4] = {0, 0, 0, 0};

   if (argc == 6) /* chip gpio1 gpio2 gpio3 gpio4 */
   {
      chip = atoi(argv[1]);
      GPIO[0] = atoi(argv[2]);
      GPIO[1] = atoi(argv[3]);
      GPIO[2] = atoi(argv[4]);
      GPIO[3] = atoi(argv[5]);
   }

   else if (argc == 5) /* gpio1 gpio2 gpio3 gpio4 (chip 0) */
   {
      chip = 0;
      GPIO[0] = atoi(argv[1]);
      GPIO[1] = atoi(argv[2]);
      GPIO[2] = atoi(argv[3]);
      GPIO[3] = atoi(argv[4]);
   }

   else
   {
      fprintf(stderr, "Usage: ./28BYJ_48 [chip] gpio1 gpio2 gpio3 gpio4\n");
      return -1;
   }

   sbc = rgpiod_start(NULL, NULL);

   if (sbc < 0)
   {
      printf("connection failed\n");
      exit(-1);
   }

   h = gpiochip_open(sbc, chip);

   if (h < 0)
   {
      fprintf(stderr, "can't open gpiochip %d (%s)\n",
         chip, lgu_error_text(h));
      return -1;
   }

   err = group_claim_output(sbc, h, 0, 4, GPIO, ZEROS);

   if (err < 0)
   {
      fprintf(stderr, "can't claim GPIO for output (%s)\n",
         lgu_error_text(err));
      return -1;
   }

   for (i=0; i<1000; i++)
   {
      group_write(sbc, h, GPIO[0],  forward(&pos), -1);
   }

   for (i=0; i<1000; i++)
   {
      group_write(sbc, h, GPIO[0],  backward(&pos), -1);
   }

   group_free(sbc, h, GPIO[0]);

   gpiochip_close(sbc, h);

   rgpiod_stop(sbc);
}

File: EXAMPLES/py_lgpio/28BYJ_48.py
#!/usr/bin/env python
"""
28BYJ_48.py
2020-11-22
Public Domain

http://abyz.me.uk/lg/py_lgpio.html

./28BYJ_48.py [chip] gpio1 gpio2 gpio3 gpio4

E.g.

./28BYJ_48.py 20 21 22 23 # gpiochip=0 gpio1=20 gpio2=21 gpio3=22 gpio4=23

./28BYJ_48.py 2 7 5 11 3  # gpiochip=2 gpio1=7 gpio2=5 gpio3=11 gpio4=3
"""

class stepper:
   """
   A class to pulse a stepper.
   """

   on = [7, 3, 11, 9, 13, 12, 14, 6]

   def __init__(self, sbc, chip, GPIO):
      """
      """
      self._sbc = sbc
      self._chip = chip
      self._leader = GPIO[0]
      self._pos = 0

      sbc.group_claim_output(chip, GPIO)

   def move(self):

      if self._pos < 0:
         self._pos = 7
      elif self._pos > 7:
         self._pos = 0

      self._sbc.group_write(self._chip, self._leader, stepper.on[self._pos])

   def forward(self):
      self._pos += 1
      self.move()

   def backward(self):
      self._pos -= 1
      self.move()

   def stop(self):
      self._sbc.group_free(self._chip, self._leader)

if __name__ == "__main__":

   import time
   import sys
   import lgpio as sbc

   if len(sys.argv) == 6: # chip gpio1 gpio2 gpio3 gpio4
      chip = int(sys.argv[1])
      gpio1 = int(sys.argv[2])
      gpio2 = int(sys.argv[3])
      gpio3 = int(sys.argv[4])
      gpio4 = int(sys.argv[5])

   elif len(sys.argv) == 5: # gpio1 gpio2 gpio3 gpio4 (chip 0)
      chip = 0
      gpio1 = int(sys.argv[1])
      gpio2 = int(sys.argv[2])
      gpio3 = int(sys.argv[3])
      gpio4 = int(sys.argv[4])

   else:
      print("Usage: ./28BYJ_48.py [chip] gpio1 gpio2 gpio3 gpio4")
      exit()

   chip = sbc.gpiochip_open(0)

   s = stepper(sbc, chip, [gpio1, gpio2, gpio3, gpio4])

   for i in range(4096):
      time.sleep(0.0015)
      s.forward()

   for i in range(4096):
      time.sleep(0.0015)
      s.backward()

   s.stop()

   sbc.gpiochip_close(chip)

